browser/ephemeral_storage/ephemeral_storage_tab_helper.cc

```diff
@@ -0,0 +1,183 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ephemeral_storage/ephemeral_storage_tab_helper.h"
+
+#include <set>
+#include <string>
+#include <utility>
+
+#include "base/feature_list.h"
+#include "base/no_destructor.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_list.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/session_storage_namespace.h"
+#include "content/public/browser/storage_partition.h"
+#include "content/public/browser/web_contents.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+#include "third_party/blink/public/common/features.h"
+
+#if defined(OS_ANDROID)
+#include "chrome/browser/android/tab_android.h"
+#include "chrome/browser/ui/android/tab_model/tab_model_list.h"
+#endif
+
+using content::BrowserContext;
+using content::NavigationHandle;
+using content::WebContents;
+
+namespace {
+
+content::SessionStorageNamespaceMap& session_storage_namespace_map() {
+  static base::NoDestructor<content::SessionStorageNamespaceMap>
+      session_storage_namespace_map;
+  return *session_storage_namespace_map.get();
+}
+
+content::SessionStorageNamespaceMap& local_storage_namespace_map() {
```

Comment:
I'm slightly concerned that this is global for all profiles. It looks like different helpers from different profiles (including incognito and TOR) will use the same namespace for a given domain. Shouldn't we keep namespaces in a keyed service or something like that?

Reply:
Great point! I'll take a look at this tomorrow.

Reply:
Okay. The new version of this branch should now key this information with the BrowserContext as well.